File: sdk/python/iotype/realtime.py
"""Realtime ASR over WebSocket.

Requires the optional dependency::

    pip install iotype-ai[realtime]

The protocol implemented here is taken from a tested browser client
(``examples/browser-asr/`` in the repository), not from prose.
"""

from __future__ import annotations

import contextlib
import json
import threading
from collections.abc import Iterator
from typing import Any, Callable, Literal

from .errors import RealtimeError

DEFAULT_WS_PATH = "/socket/realtime"

#: Audio format constants. Note that the *sample rate* is negotiated — the
#: server tells you which one to use in its authorization reply.
SAMPLE_WIDTH = 2  # PCM linear 16-bit
CHANNELS = 1      # mono

#: Frame duration used when slicing audio, in seconds. 20 ms is the value the
#: reference client uses.
FRAME_SECONDS = 0.02


class RealtimeSession:
    """A streaming speech-recognition session.

    The protocol has four steps and skipping any of them breaks the session:

    1. Send the handshake, nested inside a ``config`` object.
    2. **Wait for the reply.** It carries :attr:`sample_rate`, the rate you must
       resample audio to. Sending audio before this arrives closes the socket.
    3. Stream PCM 16-bit mono little-endian audio as binary frames, 20 ms each.
    4. Send ``{"eof": 1}`` and wait a few seconds before closing, or the final
       utterance is lost.

    Audio must never be base64-encoded.

    Example::

        with io.realtime(model="io-fa") as session:
            print("server wants", session.sample_rate, "Hz")

            threading.Thread(target=feed_audio, args=(session,), daemon=True).start()

            for event in session:
                if event["type"] == "final":
                    print(event["text"])

    Or with callbacks::

        session.run(audio_source(), on_partial=..., on_final=print)
    """

    def __init__(
        self,
        *,
        token: str,
        token_type: Literal["access_token", "flash_token"] = "access_token",
        model: Literal["io-fa", "io-en", "io-ar"] = "io-fa",
        base_url: str = "https://iotype.com",
    ) -> None:
        self.token = token
        self.token_type = token_type
        self.model = model
        self.url = base_url.replace("https://", "wss://").replace("http://", "ws://") + DEFAULT_WS_PATH

        #: Sample rate the server expects, in Hz. Populated by :meth:`connect`.
        #: **Resample your audio to this value — it is not a fixed constant.**
        self.sample_rate: int | None = None
        #: Model the server selected, echoed back in the authorization reply.
        self.negotiated_model: str | None = None

        self._ws: Any = None
        self._lock = threading.Lock()

    # ------------------------------------------------------------------ session

    def connect(self, timeout: float = 30.0) -> RealtimeSession:
        """Open the socket, send the handshake, and wait for authorization.

        Returns once the server has accepted the token. :attr:`sample_rate` is
        set before this returns.

        Raises:
            RealtimeError: if the server rejects the token or does not reply.
        """
        try:
            from websocket import create_connection  # type: ignore
        except ImportError as exc:  # pragma: no cover
            raise RealtimeError(
                "Realtime ASR needs the websocket-client package. "
                'Install it with:  pip install "iotype-ai[realtime]"'
            ) from exc

        self._ws = create_connection(self.url, timeout=timeout)

        # Step 1. The handshake must be the first message, and the three fields
        # live inside a "config" envelope.
        self._ws.send(json.dumps({
            "config": {
                "model": self.model,
                "type": self.token_type,
                "token": self.token,
            }
        }))

        # Step 2. Wait for the reply before any audio is sent.
        raw = self._ws.recv()
        try:
            reply = json.loads(raw)
        except (TypeError, json.JSONDecodeError) as exc:
            self.close()
            raise RealtimeError(f"Unparseable authorization reply: {raw!r}") from exc

        if reply.get("error"):
            self.close()
            raise RealtimeError(f"Authorization rejected: {reply['error']}")

        if reply.get("status") != "authorized":
            self.close()
            raise RealtimeError(f"Unexpected authorization reply: {reply!r}")

        self.sample_rate = reply.get("sample_rate")
        self.negotiated_model = reply.get("model")

        if not self.sample_rate:
            self.close()
            raise RealtimeError(
                "Server did not return a sample_rate. Audio cannot be sent "
                "without knowing the rate to resample to."
            )

        return self

    @property
    def frame_size(self) -> int:
        """Samples per 20 ms frame at the negotiated rate."""
        if not self.sample_rate:
            raise RealtimeError("Not connected — sample_rate is unknown.")
        return round(self.sample_rate * FRAME_SECONDS)

    def send_audio(self, chunk: bytes) -> None:
        """Send one frame of raw PCM 16-bit mono little-endian audio.

        The audio must already be at :attr:`sample_rate`.
        """
        if self._ws is None:
            raise RealtimeError("Session is not connected. Call connect() first.")
        try:
            from websocket import ABNF  # type: ignore
        except ImportError as exc:  # pragma: no cover
            raise RealtimeError("websocket-client is not installed.") from exc

        with self._lock:
            self._ws.send(chunk, opcode=ABNF.OPCODE_BINARY)

    def __iter__(self) -> Iterator[dict[str, Any]]:
        """Yield normalised events until the socket closes.

        Each event is ``{"type": "partial" | "final", "text": str}``.

        The wire format uses two different keys rather than a ``type`` field —
        ``{"partial": "..."}`` for interim text and ``{"text": "..."}`` for
        settled text. This iterator normalises both into one shape.

        ``partial`` events may be revised; render them but never persist them.
        ``final`` events are settled; persist those.
        """
        if self._ws is None:
            raise RealtimeError("Session is not connected. Call connect() first.")

        while True:
            try:
                raw = self._ws.recv()
            except Exception:  # any failure here means the socket is gone
                return
            if not raw:
                return
            if isinstance(raw, bytes):
                continue  # server sends JSON text; ignore stray binary
            try:
                data = json.loads(raw)
            except json.JSONDecodeError:
                continue

            if isinstance(data.get("partial"), str):
                yield {"type": "partial", "text": data["partial"]}
            if isinstance(data.get("text"), str):
                yield {"type": "final", "text": data["text"]}

    def run(
        self,
        audio: Iterator[bytes],
        *,
        on_partial: Callable[[str], None] | None = None,
        on_final: Callable[[str], None] | None = None,
    ) -> str:
        """Stream ``audio`` and dispatch results to callbacks.

        Returns the concatenated final transcript. The audio pump runs on a
        worker thread so a slow network cannot block your capture loop. Sends
        ``eof`` automatically once the iterator is exhausted.
        """
        if self._ws is None:
            self.connect()

        finals: list[str] = []

        def pump() -> None:
            try:
                for chunk in audio:
                    self.send_audio(chunk)
            finally:
                self.end_of_stream()

        worker = threading.Thread(target=pump, daemon=True)
        worker.start()

        for event in self:
            text = event.get("text", "")
            if event["type"] == "partial":
                if on_partial:
                    on_partial(text)
            elif text.strip():
                finals.append(text.strip())
                if on_final:
                    on_final(text)

        worker.join(timeout=5)
        return " ".join(finals)

    def end_of_stream(self) -> None:
        """Tell the server no more audio is coming and to flush its decoder.

        The last ``final`` result arrives shortly afterwards. Do not close the
        socket immediately — wait a few seconds, or iterate until it closes.
        """
        if self._ws is None:
            return
        # Best effort — the socket may already be closed by the server.
        with self._lock, contextlib.suppress(Exception):
            self._ws.send(json.dumps({"eof": 1}))

    def close(self) -> None:
        if self._ws is not None:
            try:
                self._ws.close()
            finally:
                self._ws = None

    def __enter__(self) -> RealtimeSession:
        return self.connect()

    def __exit__(self, *exc: object) -> None:
        self.close()


def float32_to_pcm16(samples: Iterator[float] | Any) -> bytes:
    """Convert normalised float samples in [-1, 1] to PCM 16-bit little-endian.

    Accepts a numpy array or any iterable of floats.
    """
    try:
        import numpy as np  # type: ignore

        if isinstance(samples, np.ndarray):
            clipped = np.clip(samples, -1.0, 1.0)
            return (clipped * 32767.0).astype("<i2").tobytes()
    except ImportError:
        pass

    import struct

    out = bytearray()
    for sample in samples:
        s = max(-1.0, min(1.0, float(sample)))
        out += struct.pack("<h", int(s * 32767))
    return bytes(out)


def resample_linear(samples: Any, input_rate: int, output_rate: int) -> Any:
    """Resample float samples with linear interpolation.

    A convenience for the common case where your capture device runs at one
    rate and the server asked for another. Requires numpy.

    For production audio quality prefer a proper resampler such as
    ``soxr`` or ``scipy.signal.resample_poly``; linear interpolation is what the
    reference browser client uses and is adequate for speech.
    """
    try:
        import numpy as np  # type: ignore
    except ImportError as exc:  # pragma: no cover
        raise RealtimeError("resample_linear requires numpy.") from exc

    if input_rate == output_rate:
        return samples

    data = np.asarray(samples, dtype=np.float32)
    if data.size == 0:
        return data

    duration = data.size / input_rate
    target = round(duration * output_rate)
    positions = np.linspace(0, data.size - 1, target, dtype=np.float64)
    return np.interp(positions, np.arange(data.size), data).astype(np.float32)

File: sdk/go/iotype/realtime.go
package iotype

import (
	"context"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// FrameSeconds is the frame duration used when slicing audio.
//
// There is deliberately no fixed sample-rate constant: the server dictates the
// rate in its authorization reply. Read RealtimeSession.SampleRate.
const FrameSeconds = 0.02

// RealtimeOptions configures a streaming ASR session.
type RealtimeOptions struct {
	Model     Model     // default ModelPersian
	Token     string    // defaults to the client's token
	TokenType TokenType // default AccessToken
	// Timeout for the authorization handshake. Default 30s.
	AuthTimeout time.Duration
}

// authResult is the server's reply to the handshake.
type authResult struct {
	Status     string `json:"status"`
	Model      Model  `json:"model"`
	SampleRate int    `json:"sample_rate"`
	Error      string `json:"error"`
}

// Event is one recognition result from the server.
//
// The wire format uses two different keys rather than a type field —
// {"partial":"..."} for interim text and {"text":"..."} for settled text.
// This type normalises both into one shape.
//
// A "partial" event is interim and may be revised — render it, never persist
// it. A "final" event is settled for that utterance and will not change —
// persist that one. A session produces many finals.
type Event struct {
	Type string // "partial" or "final"
	Text string
}

// RealtimeSession is a streaming speech-recognition session.
//
// The protocol has four steps and skipping any of them breaks the session:
//
//  1. Send the handshake, nested inside a "config" object.
//  2. Wait for the reply. It carries SampleRate — the rate to resample to.
//     Sending audio before it arrives closes the socket.
//  3. Stream PCM 16-bit mono little-endian audio as binary frames, 20 ms each.
//  4. Call Finish (or EndOfStream) before closing, or the last utterance is lost.
//
// Never base64-encode the audio.
type RealtimeSession struct {
	conn *websocket.Conn

	// SampleRate is the rate the server expects audio at, in Hz. Resample to
	// exactly this value — it is not a fixed constant across deployments.
	SampleRate int
	// NegotiatedModel is the model the server selected.
	NegotiatedModel Model

	mu        sync.Mutex
	committed strings.Builder
	partial   string
}

// Realtime opens a streaming ASR session and completes the handshake.
//
// It returns only after the server has authorized the connection, so
// SampleRate is populated on the returned session.
//
// Defaults to AccessToken because this SDK runs server-side. Never ship an
// access token to a browser or mobile client — mint a Flash Token for those
// and pass TokenType: FlashToken.
func (c *Client) Realtime(ctx context.Context, opts *RealtimeOptions) (*RealtimeSession, error) {
	model, token, tokenType := ModelPersian, c.token, AccessToken
	authTimeout := 30 * time.Second
	if opts != nil {
		if opts.Model != "" {
			model = opts.Model
		}
		if opts.Token != "" {
			token = opts.Token
		}
		if opts.TokenType != "" {
			tokenType = opts.TokenType
		}
		if opts.AuthTimeout > 0 {
			authTimeout = opts.AuthTimeout
		}
	}

	url := strings.Replace(strings.Replace(c.baseURL, "https://", "wss://", 1), "http://", "ws://", 1)
	conn, _, err := websocket.DefaultDialer.DialContext(ctx, url+"/socket/realtime", nil)
	if err != nil {
		return nil, &RealtimeError{&Error{Message: "could not open realtime socket: " + err.Error()}}
	}

	// Step 1 — handshake. Must be the first message; the fields are nested
	// inside a "config" envelope, and audio before this closes the connection.
	handshake, _ := json.Marshal(map[string]any{
		"config": map[string]string{
			"model": string(model),
			"type":  string(tokenType),
			"token": token,
		},
	})
	if err := conn.WriteMessage(websocket.TextMessage, handshake); err != nil {
		_ = conn.Close()
		return nil, &RealtimeError{&Error{Message: "handshake failed: " + err.Error()}}
	}

	// Step 2 — wait for authorization before sending any audio.
	_ = conn.SetReadDeadline(time.Now().Add(authTimeout))
	_, raw, err := conn.ReadMessage()
	_ = conn.SetReadDeadline(time.Time{})
	if err != nil {
		_ = conn.Close()
		return nil, &RealtimeError{&Error{Message: "no authorization reply: " + err.Error()}}
	}

	var auth authResult
	if err := json.Unmarshal(raw, &auth); err != nil {
		_ = conn.Close()
		return nil, &RealtimeError{&Error{Message: fmt.Sprintf("unparseable authorization reply: %s", truncate(raw, 200))}}
	}
	if auth.Error != "" {
		_ = conn.Close()
		return nil, &RealtimeError{&Error{Message: "authorization rejected: " + auth.Error}}
	}
	if auth.Status != "authorized" {
		_ = conn.Close()
		return nil, &RealtimeError{&Error{Message: fmt.Sprintf("unexpected authorization reply: %s", truncate(raw, 200))}}
	}
	if auth.SampleRate == 0 {
		_ = conn.Close()
		return nil, &RealtimeError{&Error{Message: "server returned no sample_rate; audio cannot be sent without it"}}
	}

	return &RealtimeSession{
		conn:            conn,
		SampleRate:      auth.SampleRate,
		NegotiatedModel: auth.Model,
	}, nil
}

// FrameSize returns the number of samples in a 20 ms frame at the negotiated rate.
func (s *RealtimeSession) FrameSize() int {
	return int(math.Round(float64(s.SampleRate) * FrameSeconds))
}

// SendAudio sends one frame of raw PCM 16-bit mono little-endian audio.
// The audio must already be at SampleRate.
func (s *RealtimeSession) SendAudio(chunk []byte) error {
	if s.conn == nil {
		return &RealtimeError{&Error{Message: "session is closed"}}
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.conn.WriteMessage(websocket.BinaryMessage, chunk)
}

// EndOfStream tells the server no more audio is coming and to flush its
// decoder. The last final result arrives shortly afterwards, so do not close
// the connection immediately — prefer Finish.
func (s *RealtimeSession) EndOfStream() error {
	if s.conn == nil {
		return nil
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.conn.WriteMessage(websocket.TextMessage, []byte(`{"eof":1}`))
}

// Finish sends eof, waits for the trailing result, then closes.
//
// Closing without this loses the final utterance.
func (s *RealtimeSession) Finish(wait time.Duration) (string, error) {
	if wait <= 0 {
		wait = 3 * time.Second
	}
	if err := s.EndOfStream(); err != nil {
		return s.Transcript(), err
	}
	time.Sleep(wait)
	err := s.Close()
	return s.Transcript(), err
}

// Events returns a channel of recognition results. The channel closes when the
// socket does.
func (s *RealtimeSession) Events() <-chan Event {
	out := make(chan Event, 16)

	go func() {
		defer close(out)
		for {
			msgType, raw, err := s.conn.ReadMessage()
			if err != nil {
				return
			}
			if msgType != websocket.TextMessage {
				continue
			}

			// Two shapes, told apart by which key is present.
			var msg struct {
				Partial *string `json:"partial"`
				Text    *string `json:"text"`
			}
			if json.Unmarshal(raw, &msg) != nil {
				continue
			}

			if msg.Partial != nil {
				s.mu.Lock()
				s.partial = *msg.Partial
				s.mu.Unlock()
				out <- Event{Type: "partial", Text: *msg.Partial}
			}
			if msg.Text != nil {
				text := strings.TrimSpace(*msg.Text)
				if text != "" {
					s.mu.Lock()
					s.committed.WriteString(text)
					s.committed.WriteString(" ")
					s.partial = ""
					s.mu.Unlock()
				}
				out <- Event{Type: "final", Text: text}
			}
		}
	}()

	return out
}

// Text returns what should be rendered: settled text plus the current interim
// text. Appending partials yourself produces duplicated output.
func (s *RealtimeSession) Text() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.committed.String() + s.partial
}

// Transcript returns only the settled text.
func (s *RealtimeSession) Transcript() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	return strings.TrimSpace(s.committed.String())
}

// Close tears down the session. Prefer Finish, which flushes the decoder first.
func (s *RealtimeSession) Close() error {
	if s.conn == nil {
		return nil
	}
	err := s.conn.Close()
	s.conn = nil
	return err
}

// Float32ToPCM16 converts normalised float samples in [-1, 1] to PCM 16-bit
// little-endian bytes, ready to send with SendAudio.
func Float32ToPCM16(samples []float32) []byte {
	out := make([]byte, len(samples)*2)
	for i, sample := range samples {
		s := math.Max(-1, math.Min(1, float64(sample)))
		binary.LittleEndian.PutUint16(out[i*2:], uint16(int16(s*32767)))
	}
	return out
}

// ResampleLinear resamples float samples with linear interpolation, for the
// common case where the capture device runs at one rate and the server asked
// for another.
//
// Linear interpolation is what the reference browser client uses and is
// adequate for speech. For higher audio quality use a dedicated resampler.
func ResampleLinear(samples []float32, inputRate, outputRate int) []float32 {
	if inputRate == outputRate || len(samples) == 0 {
		return samples
	}

	ratio := float64(inputRate) / float64(outputRate)
	target := int(float64(len(samples)) / ratio)
	out := make([]float32, 0, target)

	for pos := 0.0; pos+1 < float64(len(samples)); pos += ratio {
		left := int(pos)
		frac := float32(pos - float64(left))
		out = append(out, samples[left]+(samples[left+1]-samples[left])*frac)
	}
	return out
}
